onnx/helper.py

```diff
@@ -238,9 +238,15 @@ def make_tensor_value_info(name, elem_type, shape, doc_string=""):
     tensor_type_proto = value_info_proto.type.tensor_type
     tensor_type_proto.elem_type = elem_type
 
-    tensor_shape_proto = tensor_type_proto.shape.dim
+    tensor_shape_proto = tensor_type_proto.shape
+
+    # This is needed, otherwise in case of the shape is empty (which
+    # can happen when the value is a scalar), the whole "shape" field
+    # will be absent in the final proto.
+    tensor_shape_proto.dim.extend([])
```

Comment:
I don't really understand. What were the values of these variables in the bad case?

Reply:
`python -c 'import onnx; print(onnx.helper.make_tensor_value_info("Z", onnx.TensorProto.FLOAT, ()))'`

before:
```
name: "Z"
type {
  tensor_type {
    elem_type: FLOAT
  }
}
```

after:
```
name: "Z"
type {
  tensor_type {
    elem_type: FLOAT
    shape {
    }
  }
}
```

Reply:
Shall we set the shape as [0]?

Reply:
This might break the invariant num_dim == len(shape) since scalars are

Reply:
Yes, we definitely want to abide by this invariant.

Reply:
OK, I think I understand what the code is doing now. Maybe the way I would have written the comment is something like:

```
# You might think this is a no-op (extending a normal Python list by []
# certainly is), but protobuf lists work a little differently; if a field is never
# set, it is omitted from the resulting protobuf; a list that is explicitly
# set to be empty will get an (empty) entry in the protobuf. This difference
# is visible to our consumers, so make sure we emit an empty shape!
```

Reply:
@ezyang  Sure, updated the comment